qtfred/src/mission/dialogs/CampaignEditorDialogModel.cpp

```diff
@@ -0,0 +1,801 @@
+		Campaign.type = campaignType.isEmpty() ? CAMPAIGN_TYPE_SINGLE :
+												 CampaignEditorDialogModel::campaignTypes.indexOf(campaignType);
+		return {};
+	}
+	//FRED is to enforce that only on new campaigns a campaign type may be given
+	Assert(campaignType.isEmpty());
+	parse_init(false);
+	if (mission_campaign_load(qPrintable(file.replace('/',DIR_SEPARATOR_CHAR)), nullptr, 0))
+		return {};
+
+	return Campaign.filename;
+}
+
+static void initShips(const SCP_vector<ship_info>::const_iterator &s_it, CheckedDataListModel<std::ptrdiff_t> &model){
+	std::ptrdiff_t shpIdx{ std::distance(Ship_info.cbegin(), s_it) };
+	if (s_it->flags[Ship::Info_Flags::Player_ship]) {
+		model.initRow(
+				s_it->name,
+				shpIdx,
+				static_cast<bool>(Campaign.ships_allowed[static_cast<size_t>(shpIdx)]));
+}
+}
+
+static void initWeps(const SCP_vector<weapon_info>::const_iterator &w_it, CheckedDataListModel<std::ptrdiff_t> &model){
+	std::ptrdiff_t wepIdx{ std::distance(Weapon_info.cbegin(), w_it) };
+	for (const ship_info& shp: Ship_info) {
+		if (shp.flags[Ship::Info_Flags::Player_ship]
+			&& shp.allowed_weapons[static_cast<size_t>(wepIdx)]
+			&& !model.contains(w_it->name)) {
+			model.initRow(
+					w_it->name,
+					wepIdx,
+					static_cast<bool>(Campaign.weapons_allowed[static_cast<size_t>(wepIdx)]));
+}
+}
+}
+
+static SCP_vector<SCP_string> getMissions(){
+	Skip_packfile_search = 1;
+	SCP_vector<SCP_string> missions;
+	cf_get_file_list(missions, CF_TYPE_MISSIONS, "*fs2", CF_SORT_NAME);
+	Skip_packfile_search = 0;
+	return missions;
+}
+
+static bool parseMnPart(mission *mn, const char *filename){
+	try {
+		read_file_text(filename, CF_TYPE_MISSIONS);
+
+		parse_init(false);
+		Subsys_index = 0;
+		Subsys_status_size = 0;
+		if (Subsys_status != nullptr) {
+			vm_free( Subsys_status );
+			Subsys_status = nullptr;
+		}
+
+		parse_mission_info(mn, true);
+		skip_to_start_of_string("#Events");
+		Num_mission_events = 0;
+		parse_events(mn);
+		Num_goals = 0;
+		parse_goals(mn);
+	}  catch (const parse::ParseException& ) {
+		return false;
+	}
+	return true;
+}
+
+CampaignEditorDialogModel::CampaignEditorDialogModel(CampaignEditorDialog* _parent, EditorViewport* viewport, const QString &file, const QString& newCampaignType) :
+	AbstractDialogModel(_parent, viewport),
+	parent(_parent),
+	campaignFile(loadFile(file, newCampaignType)),
+	campaignType(campaignTypes[Campaign.type]),
+	initialShips(Ship_info, &initShips, this),
+	initialWeapons(Weapon_info, &initWeps, this),
+	missionData(getMissions(), &CampaignMissionData::initMissions, this),
+	campaignName(Campaign.name),		//missioncampaign.h globals
+	campaignDescr(Campaign.desc),
+	campaignTechReset(Campaign.flags & CF_CUSTOM_TECH_DATABASE)
+{
+	for (int i=0; i<Campaign.num_missions; i++) {
+		if (! missionData.contains(Campaign.missions[i].name)) {
+			mission temp{};
+			bool loaded{parseMnPart(&temp, qPrintable(Campaign.missions[i].name))};
+
+			CampaignMissionData* ptr{
```

Comment:
Where does `ptr` get deleted?

Also could you use a name like `mnDataPtr`?

Reply:
In `missionData.initRow`, missionData takes ownership of the referenced object. So ultimately, it gets destructed with the model. I'll add a comment and a longer name.